File: net/JetBrains.Serialization/src/Dataholders/AttributeInfos/MSCertificateTemplateV2AttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class MSCertificateTemplateV2AttributeInfo : AttributeInfo
{
  [JsonProperty("Identifier")] protected override TextualInfo Identifier { get; }

  [JsonProperty("Content")] private List<List<TaggedObjectInfo>> _content;

  [JsonConstructor]
  public MSCertificateTemplateV2AttributeInfo(TextualInfo identifier, List<List<TaggedObjectInfo>> content)
  {
    Identifier = identifier;
    _content = content;
  }

  public MSCertificateTemplateV2AttributeInfo(Attribute attribute)
  {
    Identifier = TextualInfo.GetInstance(attribute.AttrType);
    _content = attribute.AttrValues.ToArray().OfType<Asn1Sequence>().Select(seq =>
        seq.OfType<DerTaggedObject>()
          .Select(outer => new TaggedObjectInfo(outer.IsExplicit(), outer.TagNo, new TaggedObjectInfo(
            ((DerTaggedObject)outer.GetObject()).IsExplicit(),
            ((DerTaggedObject)outer.GetObject()).TagNo,
            TextualInfo.GetInstance(((DerTaggedObject)outer.GetObject()).GetObject()))))
          .ToList())
      .ToList();
  }

  public override Asn1Encodable GetPrimitiveContent() =>
    _content.Select(list => list.ToPrimitiveDerSequence()).ToDerSet();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/attributeInfos/MSCertificateTemplateV2AttributeInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSet
import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSequence
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.DLTaggedObject
import org.bouncycastle.asn1.cms.Attribute

// 1.3.6.1.4.1.311.2.1.12
@Serializable
data class MSCertificateTemplateV2AttributeInfo(
  override val identifier: TextualInfo,
  val content: List<List<TaggedObjectInfo>>
) : AttributeInfo() {

  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map { attributeValue ->
      (attributeValue as DLSequence).map {
        (it as DLTaggedObject).let { outer ->
          TaggedObjectInfo(
            outer.isExplicit,
            outer.tagNo,
            (outer.baseObject as DLTaggedObject).let { inner ->
              TaggedObjectInfo(
                inner.isExplicit,
                inner.tagNo,
                TextualInfo.getInstance(inner.baseObject)
              )
            }
          )
        }
      }
    }
  )

  override fun getPrimitiveContent() = content.map {
    it.toPrimitiveDLSequence()
  }.toDLSet()
}